src/circuit/circuit_frame.rs

```diff
@@ -4641,74 +4632,132 @@ pub fn comparison_helper<F: LurkField, CS: ConstraintSystem<F>>(
 }
 
 // Enforce 0 <= num < 2ˆn.
+#[allow(dead_code)]
 pub fn enforce_at_most_n_bits<F: LurkField, CS: ConstraintSystem<F>>(
     mut cs: CS,
     g: &GlobalAllocations<F>,
-    num: AllocatedNum<F>,
+    num_bits: Vec<Boolean>,
     n: usize,
 ) -> Result<(), SynthesisError> {
-    let num_bits = num.to_bits_le(&mut cs.namespace(|| "u64 remainder bit decomp"))?;
     let v = num_bits[n..255].to_vec();
-    popcount(&mut cs.namespace(|| "add all MSBs"), &v, &g.false_num);
+    popcount(&mut cs.namespace(|| "add all MSBs"), &v, &g.false_num)?;
     Ok(())
 }
 
-// Convert from num to u64.
-pub fn to_u64<F: LurkField, CS: ConstraintSystem<F>>(
+// Lurk supported uint coercion
+#[derive(Copy, Clone)]
+pub enum UnsignedInt {
+    U32,
+    U64,
+}
+
+impl UnsignedInt {
+    pub fn num_bits(&self) -> u32 {
+        match self {
+            UnsignedInt::U32 => 32,
+            UnsignedInt::U64 => 64,
+        }
+    }
+}
+
+pub fn to_unsigned_integer_helper<F: LurkField, CS: ConstraintSystem<F>>(
     mut cs: CS,
     g: &GlobalAllocations<F>,
-    maybe_u64: &AllocatedNum<F>,
+    field_elem: &AllocatedNum<F>,
+    field_bn: BigUint,
+    field_elem_bits: &[Boolean],
+    size: UnsignedInt,
 ) -> Result<AllocatedNum<F>, SynthesisError> {
-    let p64_bn = BigUint::pow(&BigUint::from_u32(2).unwrap(), 64);
-    let v = match maybe_u64.get_value() {
-        Some(v) => v,
-        None => F::zero(),
-    };
-    let field_bn = BigUint::from_bytes_le(v.to_repr().as_ref());
-
-    let (q_bn, r_bn) = field_bn.div_rem(&p64_bn);
+    let power_of_two_bn = BigUint::pow(&BigUint::from_u32(2).unwrap(), size.num_bits());
 
+    let (q_bn, r_bn) = field_bn.div_rem(&power_of_two_bn);
     let q_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "q"), q_bn)?;
     let r_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "r"), r_bn)?;
 
-    // a = b.q + r
-    let product = mul(
-        &mut cs.namespace(|| "product(q,pow(2,64))"),
-        &q_num,
-        &g.power2_64_num,
-    )?;
+    // field element = pow(2, size).q
+    let product = match size {
+        UnsignedInt::U32 => mul(
+            &mut cs.namespace(|| "product(q,pow(2,32))"),
+            &q_num,
+            &g.power2_32_num,
+        )?,
+        UnsignedInt::U64 => mul(
+            &mut cs.namespace(|| "product(q,pow(2,64))"),
+            &q_num,
+            &g.power2_64_num,
+        )?,
+    };
+
+    // field element = pow(2, size).q + r
     let sum = add(&mut cs.namespace(|| "sum remainder"), &product, &r_num)?;
-    let u64_decomp = alloc_equal(
-        &mut cs.namespace(|| "check u64 decomposition"),
-        &sum,
-        maybe_u64,
-    )?;
-    enforce_true(
-        &mut cs.namespace(|| "enforce u64 decomposition"),
-        &u64_decomp,
+    equal(&mut cs, || "check unsigned decomposition", &sum, field_elem);
```

Comment:
I think this can be optimized more. We could do it now or just leave a comment as a reminder.

Rather than allocating product and sum then enforcing equality, I think we can do all of it in a single constraint. That would go from 3 constraints to 1, a small savings — but as we add more uint sizes, it could add up. (Plus economy is its own reward.)

I think the single constraint can be something like:
`q_num * power2 = field_elem - r_num`

Reply:
done